### DISCUSSION TITLE
schema-generation using Java classes

### DISCUSSION THREAD

**Discussion by huehnerlady** [ExpediaGroup/graphql-kotlin#1149] [Q&A] [answered]
Hi,

I would like to use Java classes for types as it is a shared library with other systems.
We aere using the following dependency: `  implementation("com.expediagroup:graphql-kotlin-schema-generator:4.1.1")`

It is working in a sense that I can generate a schema and the type is used for that, even though it has a `Input` suffix which is not ideal but ok. The thing that does not work is that all the fields are non-nullable. For us it would be important though to make either part nullable, or as a compromise have all nullable. 

Is there a setting that we can use for that?

**Comment by dariuszkuc** [comment]
Hello 👋 

We rely on [KType.isMarkedNullable](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-type/is-marked-nullable.html) to determine whether a field is nullable or not. Based on the docs if you annotate your Java classes with appropriate annotations (such as `javax.annotation.Nullable` from [SpotBugs/spotbugs-annotations](https://github.com/spotbugs/spotbugs) or `org.jetbrains.annotations.Nullable` from [Jetbrains/java-annotations](https://github.com/JetBrains/java-annotations)) it *should** work.

`Input` suffix is automatically added to any input arguments to differentiate between input and output types.

---
*since we rely on reflections there might be some other issues caused by differences between Java and Kotlin

**Reply by huehnerlady** [reply]
Hi, many thanks for your answer. unfortunately with my test setup this still did not work, so I assume this is because it is an actual Javaclass and not a kotlin class. So I guess using Java classes as input types is not supported? For me it feels odd, that it does kind of work, but not quite 😅

**Reply by smyrick** [reply]
We have never tried the schema generator with Java code so this is a first. We could add support by adding logic to the generator to check for a new custom `@GraphQLNullable` annotation. This could of course throw a runtime NPE as we could start sending you null in some cases and you may not be handling it since you do not have the explicit Kotlin type system, but it would allow the generator to work with Java code.

**Reply by huehnerlady** [reply]
hmmm I guess this is a difficult one. I could maybe get a generator to use the usual nullable annotations, but a new annotation could be quite difficult. I am currently looking into generating our classes differently and so try to be able to generate kotlin classes I can then use. It is very unfortunate that KType does not seem to know for a java class to be always nullable

**Reply by huehnerlady** [reply]
I asked int he [kotlin slack workspace](https://kotlinlang.slack.com/archives/C0AVAB92L/p1621333607003800) and they basically said as JavaTypes are `flexible` they are neither nullable not non-nullable. They also gave a workaround to detect this. Maybe this is a way? 
Maybe you can have a flag in the settings saying `transformFlexibleType: [NULLABLE|NON_NULLABLE]` or something like that? 

What do you think?

**Comment by dariuszkuc** [comment, marked as answer]
@huehnerlady Java types are considered [platform types](https://kotlinlang.org/docs/java-interop.html#null-safety-and-platform-types) which yes means they are neither nullable nor non-nullable (see the link doc for details). I just tried the annotation approach specified in the [KType.isMarkedNullable](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-type/is-marked-nullable.html) and it works fine if you use annotations with **runtime** retention (i.e. `spotbugs-annotations`).

See https://github.com/dariuszkuc/graphql-kotlin-java-models which correctly generate 

```graphql
type Foo {
  getBar: String!
  getBaz: Int
}
```

from corresponding [Java model](https://github.com/dariuszkuc/graphql-kotlin-java-models/blob/master/src/main/java/com/example/model/Foo.java).

**Reply by huehnerlady** [reply]
hmm very odd, I tried that and it didn't work. maybe I had a wrong setup so I will try again with you example, thanks

**Reply by huehnerlady** [reply]
@dariuszkuc I feel rather stupid now and I don't know what I did wrong, but yourexample helped making it work, thanks :)
